### PR TITLE
Add support for compact representation of a `FermiWord` 

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
Add a function to return a `FermiWord` object from a compact string representation.

**Description of the Change:**
The `fermi/string_to_fermi_word` function is added.

**Benefits:**
Makes it easy for the users to construct a Fermi word operator with minimal work.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by soranjh** [comment]
[sc-39015]

**Review by lillian542 — APPROVED** [review]
Looks good! Left a few comments. In particular there is a small bug when using formats without an explicit minus sign if the orbital index is more than 9

File: pennylane/fermi/fermionic.py [hunk 1]
```diff
@@ -242,3 +243,44 @@ def simplify(self, tol=1e-8):
...
+    >>> string_to_fermi_word('0^ 1 0^ 1')
+    <FermiWord = '0+ 1- 0+ 1-'>
+    """
+    if fermi_string.isspace() or not fermi_string:
+        return FermiWord({})
+
+    fermi_string = re.sub(" +", " ", fermi_string).strip()
+
+    if not all(s.isdigit() or s in ["+", "-", "^", " "] for s in fermi_string):
+        raise ValueError(f"Invalid character encountered in string {fermi_string}.")
```
Comment (lillian542) [inline pennylane/fermi/fermionic.py:280]: In terms of input validation, do we also want to check that no one does this: 

`string_to_fermi_word("+0 -1 +0 -1")`

Doesn't seem like a completely unreasonable mistake, but we did tell them the format in the docstring 🤷‍♀️

File: pennylane/fermi/fermionic.py [hunk 2]
```diff
@@ -242,3 +243,44 @@ def simplify(self, tol=1e-8):
...
+
+    operators = [
+        i + "-" if len(i) == 1 else i for i in re.split(r"\s", re.sub(r"\^", "+", fermi_string))
+    ]
+
+    return FermiWord({(i, int(s[0 : len(s) - 1])): s[-1] for i, s in enumerate(operators)})
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/fermi/fermionic.py]: Corrected, thanks!

File: tests/fermi/test_fermionic.py [hunk 3]
```diff
@@ -403,3 +403,35 @@ def test_pow(self, f1, pow, result):
...
+        ("0^ 0", fw2),
+        ("0^    0", fw2),
+        ("0^ 3 0^ 4", fw3),
+        ("0+ 1", fw1),
+        ("0+  1", fw1),
+        ("0+ 0", fw2),
+        ("0+ 3 0+ 4", fw3),
+        ("", fw4),
+        (" ", fw4),
+        ("0+ 30- 0+ 400-", FermiWord({(0, 0): "+", (1, 30): "-", (2, 0): "+", (3, 400): "-"})),
```
[the remark this answers is not in the document]
Comment (soranjh) [reply tests/fermi/test_fermionic.py]: Added.

File: pennylane/fermi/fermionic.py [hunk 4]
```diff
@@ -242,3 +243,49 @@ def simplify(self, tol=1e-8):
...
+
+    >>> op1 = FermiC(0) * FermiA(1) * FermiC(2) * FermiA(3)
+    >>> op2 = string_to_fermi_word('0+ 1- 2+ 3-')
+    >>> op1 == op2
+    True
+    """
+    if fermi_string.isspace() or not fermi_string:
+        return FermiWord({})
+
+    fermi_string = re.sub(" +", " ", fermi_string).strip()
```
Comment (Jaybsoni) [inline pennylane/fermi/fermionic.py:282]: What is the purpose of this line?

**Review by Jaybsoni — APPROVED** [review]
Looks good to me!

File: pennylane/fermi/fermionic.py (hunk 4 above)
Comment (soranjh) [reply pennylane/fermi/fermionic.py]: It replaces multiple whitespaces  with a single one. Replaced it with something more clear.

**Pull request merged**